operations/rotate-linux-windows-passwords/README.md

```diff
@@ -0,0 +1,102 @@
+```
+vault server -dev -dev-root-token-id=root
+```
+
+Leave Vault running in this terminal. You can point out API actions as they are logged, such as revoked leases, etc.
+
+### Step 2: Configure a policy
+Open the Vault UI at http://127.0.0.1:8200 and create an ACL policy called `rotate-linux` with the following content:
+
+```
+path "secret/data/linux/*" {
+  capabilities = ["create", "read", "update", "list"]
+}
+```
+
+### Step 3: Generate a token
+Open a second terminal and generate a token for use in Step 5.
+```
+export VAULT_ADDR=http://127.0.0.1:8200
+vault token create -period 24h -policy rotate-linux
+```
+
+### Step 4: Configure a Linux instance
+Open a third terminal window and run the following commands. The first one creates a new Linux instance in GCP. The second command copies our password rotation script to the remote host. The third establishes an SSH connection to the remote host, while enabling a tunnel back to Vault on our local machine.
+
+```
+gcloud compute instances create linuxdemo \
+  --zone us-central1-a \
+  --image-family=ubuntu-1604-lts \
+  --image-project=ubuntu-os-cloud
+
+# You may have to wait a few seconds before running the next command. GCP is fast, but not *that* fast.
+
+gcloud compute scp --zone us-central1-a \
+  files/rotate_linux_password.sh \
+  linuxdemo:/tmp/rotate_linux_password.sh
+
+gcloud compute ssh --zone us-central1-a linuxdemo -- -p 22 -R 8200:localhost:8200
+```
+
+### Step 5: Rotate the root password
+Run these commands on the linuxdemo instance. Use the Vault token you created earlier. You do not need to install Vault on this Linux instance. The update script uses the built-in `curl` command to securely save newly generated credentials in Vault. The script requires three arguments, the username, the length of the randomly generated password, and your Vault URL.
+```
+sudo /bin/su - root
+cd /tmp
+export VAULT_ADDR=http://127.0.0.1:8200
+# Don't forget to replace with the token you created in step #3 here!
+export VAULT_TOKEN=4ebeb7f9-d691-c53f-d8d0-3c3d500ddda8
+./rotate_linux_password.sh root 12 $VAULT_ADDR
+```
+
+### Step 6: 
+Open your local Vault UI on `http://127.0.0.1:8200` and show the root password. It is stored in `secret/linux/linuxdemo/root_creds`. Run the script from step 5 again and show it update.
+
+## Optional Extras
+
+### Show older versions of the credentials
+Here you can talk about versioned KV and how older versions of the credentials are still accessible. You might need them for forensics or to know which password was used at a particular time in the past. These commands can be run from the linuxdemo instance.
+```
+export VAULT_TOKEN=root
+export VAULT_ADDR=http://127.0.0.1:8200
+apt install -y jq
+curl -X GET -H "X-Vault-Token: $VAULT_TOKEN" ${VAULT_ADDR}/v1/secret/data/linux/linuxdemo/root_creds?version=1 | jq .
+curl -X GET -H "X-Vault-Token: $VAULT_TOKEN" ${VAULT_ADDR}/v1/secret/data/linux/linuxdemo/root_creds?version=2 | jq .
+```
+
+### Do it on Windows
```

Comment:
Is there a way we can include the Windows steps in parallel and they just define which OS they want to do is for (rather than throwing it in at the end)?

Might make more sense as an update later on as it's not pressing, but potentially a better UX. If we can make each step agnostic and they just select the platform they want within the step that would be cool.

So for example, just create env vars for the paths in Vault so they set it at the top to whatever they want.

```
export PLATFORM=linux|windows

curl -X GET -H "X-Vault-Token: $VAULT_TOKEN" ${VAULT_ADDR}/v1/secret/data/${PLATFORM}/${PLATFORM}demo/root_creds?version=1 | jq .
```

Then just provide both flavors of scripts in step.

`./rotate-linux-password.sh root 12 $VAULT_ADDR`
`./rotate-windows-password.ps1 root 12 $VAULT_ADDR`

Reply:
The Windows demo would not fit within this framework.  The only reason this works on Linux is because we can forward the Vault port 8200 over the SSH connection so it is accessible via localhost on the remote machine.

So I'd recommend a completely separate howto for Windows, where one of the prerequisites is setting up a Vault server somewhere that your Windows instance can get to it.

Reply:
Makes sense - I'd just keep it as is for now then, we can tackle that at a later time if it's worth the investment.